federation/pkg/federation-controller/service/dns/dns.go

```diff
@@ -527,21 +527,27 @@ func (s *ServiceDNSController) ensureDNSRecords(clusterName string, service *v1.
 	}
 	commonPrefix := serviceName + "." + namespaceName + "." + s.federationName + ".svc"
 	// dnsNames is the path up the DNS search tree, starting at the leaf
-	dnsNames := []string{
-		strings.Join([]string{commonPrefix, zoneNames[0], regionName, s.serviceDNSSuffix}, "."), // zone level - TODO might need other zone names for multi-zone clusters
-		strings.Join([]string{commonPrefix, regionName, s.serviceDNSSuffix}, "."),               // region level, one up from zone level
-		strings.Join([]string{commonPrefix, s.serviceDNSSuffix}, "."),                           // global level, one up from region level
-		"", // nowhere to go up from global level
+	var dnsNames [][]string
+	for _, zoneName := range zoneNames {
+		zoneDNSNames := []string{
+			strings.Join([]string{commonPrefix, zoneName, regionName, s.serviceDNSSuffix}, "."), // zone level - TODO might need other zone names for multi-zone clusters
+			strings.Join([]string{commonPrefix, regionName, s.serviceDNSSuffix}, "."),           // region level, one up from zone level
+			strings.Join([]string{commonPrefix, s.serviceDNSSuffix}, "."),                       // global level, one up from region level
+			"", // nowhere to go up from global level
+		}
+		dnsNames = append(dnsNames, zoneDNSNames)
```

Comment:
You will append many duplicate entries here, because all multizone clusters always exist in the same region. You will need to remove duplicates here.

Reply:
@zihaoyu Yes.  Because of where your for loop is, on each iteration, dnsZoneNames will be:

```
[ 
   "common.zone1.region.suffix",
   "common.region.suffix",
   "common.suffix,
   ""
]
```

And hence after 3 iterations of your for loop, dnsNames will be:
```
[ 
   "common.zone1.region.suffix",
   "common.region.suffix",
   "common.suffix,
   "",

   "common.zone2.region.suffix",
   "common.region.suffix",
   "common.suffix,
   "",

   "common.zone3.region.suffix",
   "common.region.suffix",
   "common.suffix,
   ""
]
```
So your for loop needs to be around the zone entry only, so as not to duplicate all the others.

Reply:
Thanks. I will take another stab at it.